src/lua/arrow_record_batch.c

```diff
@@ -0,0 +1,76 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2024, Tarantool AUTHORS, please see AUTHORS file.
+ */
+
+#include "arrow_record_batch.h"
+
+#include "lua/utils.h"
+
+static const char *arrow_record_batch_typename = "arrow_record_batch";
```

Comment:
I don't understand the purpose of this new Lua type. It doesn't have any useful methods. It can pin quite a big chunk of non-Lua memory, which is bad for Lua GC. Why not use plain `string` or `varbinary` instead and decode the data on insertion?

Reply:
AFAIU you could:
- make the built-in msgpack decoder decode `MP_ARROW` to a raw Lua string (or varbinary);
- make `box.space.test:insert_arrow()` take a raw Lua string (or varbinary) and decode it as arrow.

There's no need for the new Lua type to achieve that. A plain Lua string (or varbinary) is better because it's accounted properly by the Lua GC, in contrast to the new Lua object type you want to introduce, which allocates arrow data from the heap.

Here's the diff to illustrate what I mean.

<details><summary>diff</summary>

```diff
diff --git a/src/box/lua/index.c b/src/box/lua/index.c
index 0d496f44dd11..5ab979a1a048 100644
--- a/src/box/lua/index.c
+++ b/src/box/lua/index.c
@@ -38,7 +38,7 @@
 #include "box/lua/misc.h"
 #include "small/region.h"
 #include "fiber.h"
-#include "lua/arrow_record_batch.h"
+#include "arrow_ipc.h"
 
 /** {{{ box.index Lua library: access to spaces and indexes
  */
@@ -430,8 +430,21 @@ lbox_insert_arrow(lua_State *L)
 	}
 	uint32_t space_id = lua_tonumber(L, 1);
 
-	struct arrow_record_batch *arrow = luaT_check_arrow_record_batch(L, 2);
-	int rc = box_insert_arrow(space_id, &arrow->array, &arrow->schema);
+	uint32_t len;
+	const char *data = luaT_tovarbinary(L, 2, &len);
+	if (data == NULL) {
+		diag_set(IllegalParams, "expected varbinary as 2 argument");
+		return luaT_error(L);
+	}
+	struct ArrowArray array;
+	struct ArrowSchema schema;
+	if (arrow_ipc_decode(&array, &schema, data, data + len) != 0)
+		return luaT_error(L);
+	int rc = box_insert_arrow(space_id, &array, &schema);
+	if (schema.release != NULL)
+		schema.release(&schema);
+	if (array.release != NULL)
+		array.release(&array);
 	if (rc != 0)
 		return luaT_error(L);
 	return 0;
diff --git a/src/lua/msgpack.c b/src/lua/msgpack.c
index b322dc0fa9bd..3b5d84777981 100644
--- a/src/lua/msgpack.c
+++ b/src/lua/msgpack.c
@@ -523,9 +523,7 @@ luamp_decode_with_ctx(struct lua_State *L, struct luaL_serializer *cfg,
 		}
 		case MP_ARROW:
 		{
-			struct arrow_record_batch *arrow =
-				luaT_new_arrow_record_batch(L);
-			arrow_unpack(data, len, &arrow->array, &arrow->schema);
+			luaT_pushvarbinary(L, *data, len);
 			return;
 		}
 		default:
```

</details>

(I'm not sure that usage of the `varbinary` type is a good idea here. Maybe, a plain string is better.)

Reply:
Thanks for the explanation, implemented this way.

Also renamed this method to `space:insert_mp_arrow()`, because probably later we'll introduce full-featured Lua wrappers for Arrow objects, that will be inserted by `space:insert_arrow()`.

@Buristan suggested to use a `string` here instead of `varbinary`.

Reply:
No need to rename. Let's still use the name `space:insert_arrow()`. If we introduce full-featured Lua wrappers for Arrow objects, we'll simply make this function take both an Arrow data string and an Arrow object (it's Lua, not C, so it's fine). Using `_mp_` in the name is actually quite confusing because this function doesn't have anything to do with MsgPack.

Reply:
OK, renamed back.